### PR TITLE
[inductor] Enable mypy checking in torch/_inductor/metrics.py

### PR DESCRIPTION
**Pull request by pgryko**
As suggested in https://github.com/pytorch/pytorch/issues/105230

Implements small fix for torch/_inductor/metrics.py

I ran into a circular import, which I handled using if TYPE_CHECKING (https://docs.python.org/3/library/typing.html#constant). 

There are then two options for describing the types, either use their class names as strings or use from future import annotations

```
If from __future__ import annotations is used, annotations are not evaluated at function definition time. 
Instead, they are stored as strings in __annotations__. 
This makes it unnecessary to use quotes around the annotation (see [PEP 563](https://peps.python.org/pep-0563/)).
```

I'm open to suggestions if it does not meet your coding guidelines

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

### CODE REVIEW

**Comment by Skylion007** [comment]
I don't think you should need the from __future__ import annotations. The if statement for the circular import should be enough.

File: torch/_inductor/metrics.py [hunk 1]
```diff
@@ -1,8 +1,15 @@
+from __future__ import annotations
```
Comment (Skylion007) [inline torch/_inductor/metrics.py:1]: This __future__ import annotatiosn should not be needed in conjuction with the circular import, can you remove and it make sure it still works?

Comment (pgryko) [reply torch/_inductor/metrics.py]: I've removed the import annotations

**Review by Skylion007 — APPROVED** [review]

**Comment by Skylion007** [comment]
You need to run `lintrunner -a` and apply the UFMT changes.

**Comment by pgryko** [comment]
@Skylion007 Linter has been run. Removing import annotations  causes tests to fail due to circular import

https://hud.pytorch.org/pr/pytorch/pytorch/105793#15261306254

If you don't want to import annotations, that I think you can describe the types as strings

https://docs.python.org/3/library/typing.html#constant

**Comment by Skylion007** [comment]
I don't think Inductor works in either case with the TypeChecking circular import trick. Can the circular import be fixed properly by moving some of the offending classes into their own file which is imported by both Python files?

**Comment by pgryko** [comment]
> I don't think Inductor works in either case with the TypeChecking circular import trick. Can the circular import be fixed properly by moving some of the offending classes into their own file which is imported by both Python files?

I re-added the from __future__ import annotations and if I'm reading this correctly, CI did pass. Subsequent CI runs failed due to cancellations
https://github.com/pytorch/pytorch/actions/runs/5699543279

On paper both import annotations and using a string replacement should work (https://docs.python.org/3/library/typing.html?highlight=type_checking#typing.TYPE_CHECKING), however when using annotations enclosed in quotes (i.e. strings), one or two benchmark tests failed (i.e. degradation in performance) (possibly due to race conditions or instability in benchmarking?). 

I'm reluctant to refactor the offending classes, as it might skew or make worse any existing non-deterministic behaviour.

**Pull request closed without merging**